src/quo2/components/input/view.cljs

```diff
@@ -0,0 +1,55 @@
+(ns quo2.components.input.view
+  (:require [quo2.components.icon :as icon]
+            [quo2.components.input.style :as style]
+            [react-native.core :as rn]
+            [reagent.core :as reagent]))
+
+(def ^:private custom-props [:type :variant :error :right-icon])
+
+(defn- base-input
+  [_]
+  (let [status (reagent/atom :default)]
+    (fn [{:keys [variant error right-icon]
+          :or   {variant :dark-blur}
+          :as   props}]
+      (let [colors-by-status (get-in style/status-colors [variant (if error :error @status)])
+            clean-props      (apply dissoc props custom-props)]
+        [rn/view
+         [rn/text-input
+          (merge {:style                  (style/input colors-by-status)
+                  :placeholder-text-color (:placeholder-color colors-by-status)
+                  :cursor-color           (:cursor-color colors-by-status)
+                  :on-focus               #(reset! status :focus)
+                  :on-blur                #(reset! status :default)}
+                 clean-props)]
+         (when-let [{:keys [on-press icon-name]} right-icon]
+           [rn/touchable-opacity
+            {:style    style/right-icon-touchable-area
+             :on-press on-press}
+            [icon/icon icon-name style/icon]])]))))
+
+(defn- password-input
+  [_]
+  (let [password-shown? (reagent/atom false)]
+    (fn [props]
+      [base-input
+       (assoc props
+              :auto-capitalize   :none
+              :auto-complete     :new-password
+              :secure-text-entry (not @password-shown?)
+              :right-icon        {:on-press  #(swap! password-shown? not)
+                                  :icon-name (if @password-shown? :i/hide :i/reveal)})])))
+
+(defn input
+  "{:type      :password
+    :variant   :dark-blur
+    ...
+    & any other TextInput available prop, such as: value, :default-value & :on-change
+    }"
```

Comment:
Hey, just want to ask a very honest question here. I constantly see outdated docstrings for the quo2 components. I have fixed a couple myself. Does anybody use these docstrings or are we just documenting for the sake of documenting?

I totally understand it's a good practice to document public APIs, but since more often than not quo2 component docstrings are outdated, don't follow a common convention/structure, and maybe nobody rely on them, then what's the point?

As of today, I have the opinion that we should eliminate all these docstrings from the components.

As a side note, Clojure can give us automatic documentation in docstrings if the functions args are spec'ed. This capability is used by Clojure core and many other libs. It's also quite easy to do, a no brainer. I have a local branch with it, and if we're interested I can refine it.

cc'ing some folks: @J-Son89, @flexsurfer, @briansztamfater, @erikseppanen, @Parveshdhull

Reply:
Please share the branch Icaro. I would much prefer to do the spec approach and have documentation in code and generated 👍

Reply:
> Does anybody use these docstrings or are we just documenting for the sake of documenting?

I recently used a button component, I didn't know how to call it, the docstring was outdated but still useful.
 

> I totally understand it's a good practice to document public APIs, but since more often than not quo2 component docstrings are outdated, don't follow a common convention/structure, and maybe nobody rely on them, then what's the point?

It saves time to have clear docstrings, when I use a component without a docstring, I search for usages for it and it takes time to realize what way I want to call it, or I need to read the code to understand what parameters are required.

> As of today, I have the opinion that we should eliminate all these docstrings from the components.

If we wouldn't have docstrings, what would you do to use a new component that you don't know how to call it? 
for example, the component added for this PR, what would you do if you need to use it and no docstring is available?

IMO reading the code to understand an API is mostly a waste of time.

> As a side note, Clojure can give us automatic documentation in docstrings if the functions args are spec'ed. This capability is used by Clojure core and many other libs. It's also quite easy to do, a no brainer. I have a local branch with it, and if we're interested I can refine it.

I think sometimes we should also specify more than "the type" of a parameter in documentation, e.g. what is a param used for, what it represents, etc.

Also, maps are usually open, for example, this input component can receive any TextInput available prop, I'd say spec'ed documentation is not enough.

What happens when the spec is not satisfied? Is spec checking disabled in production?

I'm curious about your branch, please share :)

And finally, I used to "document" component params by using destructuring (even if the resulting vars are not directly used in the function's body) and add a docstring only when it is required to explain something else.

Reply:
> As a side note, Clojure can give us automatic documentation in docstrings if the functions args are spec'ed. This capability is used by Clojure core and many other libs. It's also quite easy to do, a no brainer. I have a local branch with it, and if we're interested I can refine it.

Quite interesting, I like this approach and would love to see how it works, the main reason is maintainability and simplicity. As you say @ilmotta, I usually not rely on these type of documentation, but is nice to have it and for some devs it might be useful. If we can standarize and automatize this type of documentation would be great. Also really appreciate @ulisesmac effort to write them and forcing us to have this discussion :)

Reply:
Fair enough @ulisesmac ;) I'm not arguing against docstrings, that would be insane, just against outdated and incomplete docstrings, but if they provide value as I'm seeing from the comments, I'm fine with that. That was my main question in this thread. Thank you for the detailed answers and questions.

> Please share the branch Icaro. I would much prefer to do the spec approach and have documentation in code and generated +1

@J-Son89, I'll share a branch, but since most of us never used much of spec, I'll try to make it more comprehensible/useful before sharing.

> If we can standarize and automatize this type of documentation would be great.

@briansztamfater, there's no free lunch unfortunately, specs need to be designed and cared for :baby_bottle: Just like good docstrings don't come for free. The difference would be that specs can be executed during runtime and validate function input/output for, something these literal maps in quo2 docstrings can't do. Just to reiterate, specs would be complimentary to docstrings, they wouldn't necessarily be replacements.

> If we wouldn't have docstrings, what would you do to use a new component that you don't know how to call it?

I do the business as usual, read the function signature at the modeline in Emacs (think of it as a tooltip in VSCode or IntelliJ). In the function signature the names of the arguments and destructuring of keys is often enough to tell me what needs to be passed. `quo2` component arguments tend to be quite repetitive, like `override-theme`, `color`, `icon`, etc. I quickly developed a mental model of the argument types.

But of course, this is my personal approach to functions without docstrings. This is fairly personal, just answering your question.

> IMO reading the code to understand an API is mostly a waste of time.

To each their own. In a perfect world where everything is well documented, maybe this perspective holds, but in the real world, documentation is often not enough, so reading the source code is hardly a waste of time.

> I think sometimes we should also specify more than "the type" of a parameter in documentation, e.g. what is a param used for, what it represents, etc.

Agreed. What the specs could replace are the definition of argument types and their shape (data structure), not the explanation of what they do.

> What happens when the spec is not satisfied? Is spec checking disabled in production?

It depends, `spec` can have numerous usages. By the word *checking* I take you're referring to *instrumentation*? Instrumentation is pretty much always disabled in production, because there's no point in verifying compiled code. The compiled code is simply assumed to be correct.

But spec has many other usages, like conforming data (awesome for macros), generative testing, documentation, validation, among other things.

A superior alternative to spec in my experience is Malli, which I extensively used in the past. But that's a tangent for now.

> And finally, I used to "document" component params by using destructuring (even if the resulting vars are not directly used in the function's body) and add a docstring only when it is required to explain something else.

Great approach, and I see that we often apply this practice in quo2 components. It's one of the reasons I (generally) don't feel the need for much documentation in quo2 components.

> Also, maps are usually open, for example, this input component can receive any TextInput available prop, I'd say spec'ed documentation is not enough.

I don't understand this point Ulises. Clojure is entirely based on open maps, Rich H. extensively talked about this. Instead of adding `& any other TextInput available prop, such as: value, :default-value & :on-change`, with specs you can just define a new spec for the supported TextInput properties if you wanted, that'd be even better.

Reply:
> I'll share a branch, but since most of us never used much of spec, I'll try to make it more comprehensible/useful before sharing.

Is it possible to write them using mali? :eyes:  I was used to use spec, but tried [malli](https://github.com/metosin/malli) and liked it, is more complete IMO.

>  so reading the source code is hardly a waste of time.

I agree! just wanted to say that if you want to use a function, reading the code to understand which vars are being used is not the best approach. Of course, I understand this is the real world and not everything will be as we wish! :)

> Clojure is entirely based on open maps, Rich H. extensively talked about this. Instead of adding & any other TextInput available prop, such as: value, :default-value & :on-change, with specs you can just define a new spec for the supported TextInput properties if you wanted, that'd be even better.

It would be great for documentation, but it would take time to write specs for react native components and, as we upgrade RN versions, they'd need to be maintained.

I'm not against maps, in fact I really like them, but trying to spec them, when they're tied to an external API (such as RN components) could be hard.

Thanks for your answers! :)

Reply:
Interesting discussion!

I think the things that add the most value are the things that can be routinely executed, like tests (speaking of which, tests can be useful for understanding an API). In general, docstrings tend to become outdated, because the original author cared the most about it, and subsequent developers often don't grok the logic deeply enough (and maybe don't want to change the author's original wording, because sometimes you have to re-write the whole docstring for it to continue to be accurate) to maintain it adequately. And I'd say it's harder to find developers that care about maintaining docstrings that those that care about maintaining tests.

I was on a project once that required documentation, and the docstrings were leveraged to generate that documentation.

I've used spec and schema but have heard good things about malli!